Controller/12511-Added Test and Storybook files for Stimulus DrilldownController

Srishti-j18 opened the pull request:
fixed #12511

- [X] `Unit Test`: Added  test.js for the `DrilldownController`.
- [x] `Storybook Integration`: Added a Storybook file for the `DrilldownController`.
- [X] `Documentation Update`: Included an `@example` in the JSDoc for better clarity and usage guidance.
- [X] Updated the other controller imports in the file as `type` imports, as they are not used for any actual code.

lb- commented on package.json:
```diff
@@ -154,4 +154,4 @@
     "build-storybook": "storybook build -c client/storybook"
   }
-}
+}
```
It looks like your editing environment may be set up with a different end of file line break.  

Maybe check - here's some useful docs. 
https://docs.wagtail.org/en/stable/contributing/developing.html#development-on-windows

lb- commented on package-lock.json:
```diff
@@ -32860,7 +32863,9 @@
     "http-proxy-middleware": {
-      "version": "2.0.3",
+      "version": "2.0.7",
+      "resolved": "https://registry.npmjs.org/http-proxy-middleware/-/http-proxy-middleware-2.0.7.tgz",
```
Please avoid changes to the package lock, this update may be useful but should be a standalone PR. 

In the future it's best to use `npm ci` instead of `npm install` when running locally. This way the package lock file won't change. 

See https://docs.wagtail.org/en/stable/contributing/developing.html#setting-up-the-wagtail-codebase

lb- commented on client/src/controllers/DrilldownController.ts:
```diff
@@ -63,6 +74,7 @@ export class DrilldownController extends Controller<HTMLElement> {
    * are counted.
    */
   updateCount() {
+    if (!this.countAttrValue) return; // Exit if countAttrValue is empty
```
Could you please explain this change. Unless we've found a bug it's best to avoid functional code changes when just adding tests.

lb- commented on client/src/controllers/DrilldownController.ts:
```diff
@@ -1,10 +1,21 @@
  * state management for listing filters.
+ * 
+ * @example
+ * <section ><div data-controller="w-drilldown">
```
```suggestion
 * ```html
 * <section ><div data-controller="w-drilldown">
```

Can you wrap this in three backticks, like markdown. 

We don't do this consistently yet but it's nice for new examples.

lb- commented on client/src/controllers/DrilldownController.test.js:
```diff
@@ -0,0 +1,90 @@
+            expect(toggleButton).not.toBeNull();
+            expect(menu).not.toBeNull();
+
+            if (!toggleButton || !menu) return;
```
```suggestion
```
Best to avoid logic like this in tests, it could mean the tests pass when something is not right. 

If the elements do not exist in the test it means the html or selectors are wrong. 

For unit tests, we want it to fail loudly, your expect statements above will be enough to ensure the test won't continue if the elements are missing.

lb- commented on client/src/controllers/DrilldownController.test.js:
```diff
@@ -0,0 +1,90 @@
+            toggleButton.click();
+            await jest.runAllTimers();
+
+            expect(menu.hidden).toBe(true);
```
It looks like that after your click the expectations are then same as before, maybe try out the actual behaviour in the browser and see what's meant to happen.

lb- commented on client/src/controllers/DrilldownController.test.js:
```diff
@@ -0,0 +1,90 @@
+            events[name] = [];
+        });
+    });
+    // ------------------------------------
```
```suggestion

```

No need for comments like this, that's what the `describe` blocks are for.

lb- commented on client/src/controllers/DrilldownController.test.js:
```diff
@@ -0,0 +1,90 @@
+    // ------------------------------------
+    describe('when connecting to the DOM', () => {
+        it('opens and closes the menu when toggle is clicked', async () => {
+            const [controller] = await setup();
```
Unless you're actually checking something with the controller, you probably can just call setup without a variable assignment. 


```suggestion
            await setup();
```

lb- reviewed (commented):
@Srishti-j18 thank you for putting up a draft for feedback.  

I've got some comments that might help and I'll try to also give you some general guidance. 

I'd encourage you to look at where this controller is used in the Wagtail code, it will give you a sense of the html needed and also how it interacts with other controllers through the Outlets feature.

lb- commented:
Some generic comments/ suggestions added to the issue. 
https://github.com/wagtail/wagtail/issues/12511#issuecomment-2453229083

Srishti-j18 commented:
Hi LB,
I have added the changes.

lb- commented:
Thanks @Srishti-j18  I have flagged this for a new review.

lb- commented on client/src/controllers/DrilldownController.stories.js:
```diff
@@ -0,0 +1,95 @@
+  },
+};
+
+// Separate each controller registration
```
```suggestion
```

No need for this comment.

Srishti-j18 replied:
ok, I removed it.

lb- commented on client/src/controllers/DrilldownController.test.js:
```diff
@@ -0,0 +1,214 @@
+  let actionElement;
+
+  const setup = async (
+    html = `<section>
```
If possible, are you able to clean up the HTML used in these test snippets.

1. Try to focus on the minimum DOM needed (e.g. remove most classes, unless they are used for testing or to help differentiate elements).
2. Change all nested spacing to just two spaces, it makes the HTML much more readable
3. try to align the opening/closing tags so that it's easier to spot any DOM issues
4. A reminder that `className` will not work here, it's `class` when working with HTML

Srishti-j18 replied:
Okay, I will clean up and format the HTML as you mentioned. Regarding your fourth point, my bad,I had previously written `class` instead of` className`. When I ran `npm run lint` to check for issues, it suggested using `className` instead of `class`, but this only applies to stories.js where I used React. I'll make the change.

lb- commented on client/src/controllers/DrilldownController.test.js:
```diff
@@ -0,0 +1,214 @@
+  let application;
+  let errors = [];
+  let controller;
+  let actionElement;
```
I would avoid setting controller/actionelement as globals, instead pull them in as needed in each unit test.

It's not ideal to be testing directly against the `controller` instance, this can cause brittle unit tests that rely on the inner implementation of the controller and not the events/DOM that should be used.

I would encourage you to try to review your tests and instead of doing things like `controller.close()` set an element action that will call this instead `data-action="w-dropdown:hidden->w-drilldown#delayedClose"`.

lb- commented on client/src/controllers/DrilldownController.test.js:
```diff
@@ -0,0 +1,214 @@
+      );
+      toggleButton.click();
+
+      controller.close();
```
As noted above, try to avoid directly interacting with the controller instance, instead either add more data actions or see the behaviour as it's used.

Such as in `wagtail/admin/templates/wagtailadmin/shared/headers/_filters.html`

```
<button class="w-drilldown__back" type="button" aria-label="{% trans 'Back' %}" data-action="click->w-drilldown#close">
```

Here the button's click event can trigger it.

lb- commented on client/src/controllers/DrilldownController.test.js:
```diff
@@ -0,0 +1,214 @@
+      controller.close();
+
+      const submenu = document.querySelector('#drilldown-field-0');
+      expect(controller.activeSubmenuValue).toBe('');
```
Try to avoid reading values directly from the controller.

Either test the desired behaviour (as in no sub-menu, the previous one being closed)

OR if really needed, you can check the attribute in the DOM `data-w-drilldown-active-submenu-value`.

Here, though it looks like you do not need this line here as you already are checking the DOM in the lines below.

lb- commented on client/src/controllers/DrilldownController.test.js:
```diff
@@ -0,0 +1,214 @@
+
+  describe('DrilldownController delayedClose method', () => {
+    it('should call close after 200ms delay', () => {
+      jest.spyOn(controller, 'close');
```
As above, try to avoid spys, instead check for the difference in the DOM itself.

lb- commented on client/src/controllers/DrilldownController.test.js:
```diff
@@ -0,0 +1,214 @@
+      controller.wActionOutletConnected(null, actionElement);
+      controller.wActionOutletDisconnected(null, actionElement);
+      actionElement.click();
+      expect(controller.open).not.toHaveBeenCalled();
```
This is a tricky one, I would recommend you simply delete the DOM elements (the action outlets), run all timers async (await this) and then try to check for events you can listen to.

This, as written, is not really testing the behaviour but rather 'digging into' the implementation, which is best to avoid.

lb- commented on client/src/controllers/DrilldownController.ts:
````diff
@@ -1,10 +1,62 @@
+ * @example
+ * ```html
+ *    <section>
+            <div id="filters-drilldown" className="w-drilldown" data-controller="w-drilldown" data-action="w-swap:success@document->w-drilldown#updateCount w-dropdown:hide->w-drilldown#delayedClose w-dropdown:clickaway->w-drilldown#preventOutletClickaway"
````
Please include the `*` and try to nest these with two spaces only, it's much easier to read.
Please also remove all but the bare minimum DOM that's needed to show an example. Try to find a minimal reproducible usage, test it yourself (in storybook or in the admin) and then use that as an example.

We should not need any classes (or just those that help differentiate elements), and a reminder that it's `class` not `className` in HTML.

```suggestion
 *   <div id="filters-drilldown" className="w-drilldown" data-controller="w-drilldown" data-action="w-swap:success@document->w-drilldown#updateCount w-dropdown:hide->w-drilldown#delayedClose w-dropdown:clickaway->w-drilldown#preventOutletClickaway"
```

lb- reviewed (changes requested):
Thank you @Srishti-j18  this is an amazing start, you have really thought through the controller and how it works.

I have the next round of suggestions ready, it will help get this PR close to a full review.

Mostly it's formatting & cleaning up the HTML used.

The main current issue though is that the tests need to try to test 'behaviour' (e.g. clicking/events causes something to change in the DOM), but some of your tests test the 'implementation' (requiring knowledge that there's a close method for example).

I know this may seem trivial (especially as Stimulus actions have to have knowledge of the methods to use `w-drilldown#close`), but it's an important distinction and helps make sure the tests can be less fragile if the inner workings of the controller get refactored.

Another way to think about this is, how would you write the test if you were not able to access the controller instance in the unit test.

See how you go, the disconnected outlet event listener removal may be a bit tricky but I think you are pretty close for the others.

Srishti-j18 commented:
Thank you so much for your review, LB. I really appreciate your feedback and always learn something new from your comments. I’ll make sure to make the suggested changes in the upcoming week.

Srishti-j18 commented:
Hi LB,
I made changes.. minimized, and cleaned up the HTML in every file, including Storybook also. Now storybook looks like this(similar to previous but with only useful classes and attributes):
https://github.com/user-attachments/assets/3e657b95-0e04-43e6-97a6-9958594bd431 

And I modified the tests, but I have one test that is not passing for the `delayedClose` method. I encountered some difficulties with it because the `delayedClose` method (as seen here: https://github.com/wagtail/wagtail/blob/7566fb84e0709b9dd1139161c244ea5c49d3b4c0/client/src/controllers/DrilldownController.ts#L79C1-L99C1) does not seem to modify the `data-w-drilldown-active` attribute when` close()` is called.
I wrote a test for it, but it's not passing. I have checked several things, but I’m still not sure why it's not working.

Could you please help me?

lb- commented on client/src/controllers/DrilldownController.test.js:
```diff
@@ -0,0 +1,201 @@
+      delayCloseButton.click();
+      expect(submenu.getAttribute('data-w-drilldown-active')).toBe('true');
+
+      jest.advanceTimersByTime(200);
```
You may need to use an async version of this, because when attributes update, this is because some promises need to resolve.

You will first need to ensure this test supports async.

```
    it('should call close after 200ms delay', async () => {
```


Then use the async version of the timer.

```suggestion
      await jest.advanceTimersByTimeAsync(200);
```

Srishti-j18 replied:
I did this, but the test still did not pass.
![image](https://github.com/user-attachments/assets/a5b52473-2208-452c-a879-a222a187640a)

Srishti-j18 replied:
I have become tired of trying to check why it is failing. I found that this test is not actually calling the `delayedClose() `method because it's not printing the message I have written in the method to check if it's calling the delayed method or not. But I can't find the reason yet. I used `data-action="w-dropdown:hidden->w-drilldown#delayedClose"` as suggested in the comment of the method, but it's still not working. As you can see in the attachment also-

![image](https://github.com/user-attachments/assets/3b10dc4b-97ba-48e6-9fff-8c87d2526d37)
![image](https://github.com/user-attachments/assets/f83f48f7-549a-4ee0-a4c2-669f5740605d)
![image](https://github.com/user-attachments/assets/8d3c66c5-dd7c-4bd5-8dd0-f4ae77d72c56)

Srishti-j18 replied:
Now , I think,I should throw my laptop and sleep....only this can fix me now 🥲

lb- replied:
I feel like this also sometimes. I'll try to take a look soon and see if I can help.

lb- commented on client/src/controllers/DrilldownController.test.js:
```diff
@@ -0,0 +1,201 @@
+
+  describe('DrilldownController - Action Outlet Connection/Disconnection', () => {
+    let actionElement;
+    let controller;
```
I am sure you are on it, but this test block will need to be updated to avoid accessing the controller directly.

For the disconnection, you could just use the `events` and make sure no events are fired after removal?

Srishti-j18 replied:
Yes, After connecting, Ensure the event listener is added correctly. And after disconnection, Ensure that nothing happens when you trigger the event (no additional actions are triggered after removing the element).

lb- commented on client/src/controllers/DrilldownController.test.js:
```diff
@@ -0,0 +1,201 @@
+      const toggleButton = document.querySelector(
+        '[data-w-drilldown-target="toggle"]',
+      );
+      toggleButton.click();
```
You probably need an assertion after this first click, something like 

```suggestion
      toggleButton.click();
      expect(submenu.hidden).toBeFalsy();
```

Otherwise this is potentially testing the wrong behaviour.

lb- replied:
OK, down here, there's a similar issue to the one you were having above. The actual 'outside click' is sorted by the other dropdown (not drilldown) controller, so this unit test just needs to test the firing of that event.

We want to test the `preventOutletClickaway` method here, so we look for that action and see `w-dropdown:clickaway->w-drilldown#preventOutletClickaway` is used. 

This means the event listener is attached to the event with the name`w-dropdown:clickaway`. A reminder that these names are nothing special, they could be 'foo', 'custom:event' or even 'click'. This is how actions work https://stimulus.hotwired.dev/reference/actions

So, below, is a functional test that's testing the firing of the 'w-dropdown:clickaway' event, but rather than checking it's closed, we actually have to look at what the `preventOutletClickaway` is doing.

You can see in the method it sometimes prevents the default behaviour of this event, which is what we want to emulate in the test.



```js

  it('should close the menu when clicking outside the controlled area', async () => {
    const otherTrigger = document.createElement('button');
    otherTrigger.id = 'other-trigger';
    otherTrigger.type = 'button';
    otherTrigger.setAttribute('aria-controls', 'drilldown-field-0');
    document.body.appendChild(otherTrigger);

    const drilldownContainer = document.getElementById('filters-drilldown');
    const toggleButton = document.getElementById('drilldown-toggle');
    const submenu = document.getElementById('drilldown-field-0');

    // Toggle the submenu & confirm it's open
    toggleButton.click();
    await Promise.resolve();
    expect(submenu.hidden).toBeFalsy();

    // Click outside the controlled area using the dropdown controller's event
    let event = new CustomEvent('w-dropdown:clickaway', {
      detail: { target: document.getElementById('other-trigger') },
    });
    event.preventDefault = jest.fn();
    drilldownContainer.dispatchEvent(event);
    await jest.runAllTimersAsync();

    // Confirm the event default was prevented
    expect(event.preventDefault).toHaveBeenCalled();

    // Confirm the event default is not prevented if the target does not match
    event = new CustomEvent('w-dropdown:clickaway', {
      detail: { target: document.getElementById('filters-drilldown') },
    });
    event.preventDefault = jest.fn();
    drilldownContainer.dispatchEvent(event);
    await jest.runAllTimersAsync();

    // Confirm the event default was not prevented
    expect(event.preventDefault).not.toHaveBeenCalled();
  });
```

lb- commented on client/src/controllers/DrilldownController.test.js:
```diff
@@ -0,0 +1,201 @@
+        '[data-w-drilldown-target="toggle"]',
+      );
+      toggleButton.click();
+
```
When doing a test like this, think about structuring your tests with Arrange, Act, Assert.
https://docs.telerik.com/devtools/justmock/basic-usage/arrange-act-assert

In it's current form it's not clear **why** the test is running this first `click()`. I assume you want to 'open' the drilldown and test that closing it works.

In which case, this is really part of your 'arranging', you are getting the DOM into a place so you can check something else.

You can probably add a new assertion under this first click, checking that the thing is open (or ready), this way you know that the 'arrange' has been done.

Sometimes it helps to even have comments (but remove them before the commit).

```
    it('closes the submenu and clears active submenu value when close action is triggered', () => {
      // arrange
      const toggleButton = document.querySelector(
        '[data-w-drilldown-target="toggle"]',
      );
      toggleButton.click();
     // add an assertion here that confirms the DOM is correctly arranged, I assume
    expect(submenu.hidden).toBeFalsy();

      const closeButton = document.querySelector(
        '[data-action="click->w-drilldown#close"]',
      );

      // act
      closeButton.click();

      const submenu = document.querySelector('#drilldown-field-0');

      // assert
      expect(toggleButton.getAttribute('aria-expanded')).toBe('false');
      expect(submenu.hidden).toBeTruthy();
    });
```

lb- replied:
Here's another link with Jest examples.
https://robertmarshall.dev/blog/arrange-act-and-assert-pattern-the-three-as-of-unit-testing/

lb- commented on client/src/controllers/DrilldownController.test.js:
```diff
@@ -0,0 +1,211 @@
+  let application;
+  let errors = [];
+
+  const setup = async (
```
OK, I think the HTML still needs some work, it's super hard to parse/read in it's current format & nesting, here's a refined version that will be used for the other fixes below.

Observe we are now making some (not all) of the attributes drop to new lines, we have added some `id` attributes that make it easier to grab elements and have used a bit of nested JS to make the `data-action` items easier to read.

```js
  const setup = async (
    html = `
    <section>
      <div
        id="filters-drilldown"
        class="w-drilldown"
        data-controller="w-drilldown"
        data-action='${[
          'w-swap:success@document->w-drilldown#updateCount',
          'w-dropdown:hide->w-drilldown#delayedClose',
          'w-dropdown:clickaway->w-drilldown#preventOutletClickaway',
        ].join(' ')}'
        data-w-drilldown-count-attr-value="data-w-active-filter-id"
      >
        <span class="w-drilldown__count" data-w-drilldown-target="count"></span>
        <button type="button">Show Filters</button>
        <div class="w-drilldown__menu" data-w-drilldown-target="menu" hidden>
          <h2>Filter by</h2>
          <button
            id="drilldown-toggle"
            type="button"
            aria-expanded="false"
            aria-controls="drilldown-field-0"
            data-action="click->w-drilldown#open"
            data-w-drilldown-target="toggle"
          >
            Field 1
          </button>
        </div>
        <div class="w-drilldown__submenu" id="drilldown-field-0" hidden tabIndex="-1">
          <button class="w-drilldown__back" type="button" data-action="click->w-drilldown#close">Back</button>
        </div>
      </div>
      <button type="button" id="action-element" data-controller="w-action">Action</button>
    </section>`,
    identifier = 'w-drilldown',
  ) => {
    document.body.innerHTML = `<main>${html}</main>`;
```

lb- commented on client/src/controllers/DrilldownController.test.js:
```diff
@@ -0,0 +1,211 @@
+
+  describe('DrilldownController open/close actions', () => {
+    it('opens the submenu and sets active submenu value when toggle is clicked', async () => {
+      const toggleButton = document.querySelector(
+        '[data-w-drilldown-target="toggle"]',
+      );
+      const submenu = document.querySelector('#drilldown-field-0');
```
We can now more easily use `getElementById` which is a bit nicer then `querySelector` and makes the tests a bit easier to read.

I would suggest you update all tests to follow this approach when getting the toggle / submenu.

```suggestion
      const toggleButton = document.getElementById('drilldown-toggle');
      const submenu = document.getElementById('drilldown-field-0');
```

lb- commented on client/src/controllers/DrilldownController.test.js:
```diff
@@ -0,0 +1,211 @@
+    });
+
+    it('closes the submenu and clears active submenu value when close action is triggered', async () => {
+      const toggleButton = document.querySelector(
+        '[data-w-drilldown-target="toggle"]',
+      );
```
```suggestion
      const toggleButton = document.getElementById('drilldown-toggle');
```

lb- commented on client/src/controllers/DrilldownController.test.js:
```diff
@@ -0,0 +1,211 @@
+    });
+  });
+
+  describe('DrilldownController delayedClose method', () => {
+    it('should call close after 200ms delay', async () => {
+      document.body.innerHTML = `
+      <div id="drilldown-field-0">
+        <button data-action="w-dropdown:hide->w-drilldown#delayedClose">Close</button>
+        <div id="${this.activeSubmenuValue}" hidden>Submenu content</div>
+      </div>`;
+
+      const submenu = document.querySelector(`#${this.activeSubmenuValue}`);
+      const delayCloseButton = document.querySelector(
+        '[data-action="w-dropdown:hide->w-drilldown#delayedClose"]',
+      );
+
+      await jest.runAllTimersAsync();
+      submenu.hidden = false;
+      expect(delayCloseButton).not.toBeNull();
+      expect(submenu.hidden).toBe(false);
+
+      delayCloseButton.click();
+      await jest.runAllTimersAsync();
+      expect(submenu.hidden).toBe(false);
+
+      await jest.advanceTimersByTimeAsync(200);
+      expect(submenu.hidden).toBe(true);
+    });
+  });
```
here's a functional test, but let's go through what changed.

1. Avoid too much nesting of `describe`, it makes the tests harder to read
2. Avoid re-adding the words 'Drilldown controller' within each describe/test block, this is not necessary, Jest will append this in console reporting and snapshots for us, no need to add more work for yourself.
3. Following the AAA pattern, we first arrange our code, making sure the drilldown submenu is open.
4. Then we dispatch the event `w-dropdown:hide`, this is really important, it looks like you were getting a bit confused between what the toggle button does and when the delayed close is used. If you look at your actions there is `w-dropdown:hide->w-drilldown#delayedClose`. This says, when the event is fired from within (or on) the element with the action, and that event is `w-dropdown:hide`, then call the method `delayedClose` on the drilldown controller. The similarity of names is confusing here but basically the two controllers are interacting, but we just need to 'mock' the event with the name `w-dropdown:hide`.
5. Finally, we first run the timers for 180ms, check nothing has happened, then another 20ms before asserting the final desired behaviour.

```suggestion

  it('should have the ability to be closed with a delay', async () => {
    const drilldownContainer = document.getElementById('filters-drilldown');
    const submenu = document.querySelector('.w-drilldown__submenu');

    // Toggle the submenu & confirm it's open
    document.getElementById('drilldown-toggle').click();
    await Promise.resolve();
    expect(submenu.hidden).toBe(false);

    // Activate the delayed close event
    drilldownContainer.dispatchEvent(new Event('w-dropdown:hide'));

    // confirm the submenu is visible only after 200ms
    await jest.advanceTimersByTimeAsync(180);
    expect(submenu.hidden).toBe(false);

    await jest.advanceTimersByTimeAsync(20);
    expect(submenu.hidden).toBe(true);
  });
```

lb- reviewed (commented):
@Srishti-j18  here's some suggestions, I hope this helps.

Srishti-j18 commented:
Thank you for helping me, LB. I really appreciate that you took the time and effort to review my code again and again. Through your comments, I realized several areas where I made mistakes and learned some important practices:
Writing HTML/tests in a clear and readable way for everyone.
Avoiding over-nesting in `describe `blocks .
Following the AAA (Arrange, Act, Assert) pattern for better test structure.
The biggest correction for me was identifying the confusion I had when using `dropdown:hide->w-drilldown#delayedClose `instead of `w-dropdown:hide`.  etc..

I think I will never forget this issue and your help. :)

lb- commented:
Good on you @Srishti-j18 - I'll take another look soon.  

It's a tricky thing to learn a few things at once but well done on pushing through.

lb- reviewed (approved):
Thank you @Srishti-j18  this is looking good now.

I will clean up the commits and fix up the last few nitpicky items;

1. Refine the formatting of the `example` HTML
2. Simplify the Storybook story to just focus on the drilldown only
3. Remove the `events` names, these are not needed as the controller does not dispatch any events.
4. Add JSDoc for some of the class attributes
